accepted/file-enumeration.md

```diff
@@ -0,0 +1,248 @@
+    - File size
+2. Result transforms can be of any desired output type
+	- Like Linq Select(), but keeps FileData on the stack
+3. API minimizes allocations
+4. API is cross platform abstract
+3. We provide common filters and transforms
+	- To file/directory name
+	- To full path
+	- To File/Directory/FileSystemInfo
+	- DOS style filters (Legacy- `*/?` with DOS semantics, e.g. `*.` is all files without an extension)   
+	- Simple Regex filter
+	- Simpler globbing (`*/?` without DOS style variants)
+	- Set of extensions (`*.c`, `*.cpp`, `*.cc`, `*.cxx`, etc.)
+4. Recursive behavior is configurable
+	- On/Off
+	- Predicate based on FileData
+5. Can avoid throwing access denied exceptions
+
+### Non-Goals
+
+1. API will not expose platform specific data
+3. Error handling configuration is fully customizable
+
+## Design
+
+### Proposed API surface
+
+``` C#
+namespace System.IO
+{
+    /// <summary>
+    /// Delegate for filtering out find results.
+    /// </summary>
+    public delegate bool FindPredicate<TState>(ref RawFindData findData, TState state);
+
+    /// <summary>
+    /// Delegate for transforming raw find data into a result.
+    /// </summary>
+    public delegate TResult FindTransform<TResult, TState>(ref RawFindData findData, TState state);
+
+    [Flags]
+    public enum FindOptions
+	{
+        None = 0x0,
+
+        // Enumerate subdirectories
+        Recurse = 0x1,
+
+        // Skip files/directories when access is denied
+        IgnoreAccessDenied = 0x2,
+
+        // Future: Add flags for tracking cycles, etc. 
+	}
+
+    public class FindEnumerable<TResult, TState> : CriticalFinalizerObject, IEnumerable<TResult>, IEnumerator<TResult>
+    {
+        public FindEnumerable(
+            string directory,
+            FindTransform<TResult, TState> transform,
+            FindPredicate<TState> predicate,
+            // Only used if FindOptions.Recurse is set. Default is to always recurse.
+            FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None)
+    }
+
+    public static class Directory
+    {
+        public static IEnumerable<TResult> Enumerate<TResult, TState>(
+            string path,
+            FindTransform<TResult, TState> transform,
+            FindPredicate<TState> predicate,
+			FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None);
+    }
+
+    public class DirectoryInfo
+    {
+        public static IEnumerable<TResult> Enumerate<TResult, TState>(
+            FindTransform<TResult, TState> transform,
+            FindPredicate<TState> predicate,
+			FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None);
+    }
+
+    /// <summary>
+    /// Used for processing and filtering find results.
+    /// </summary>
+    public ref struct RawFindData
```

Comment:
Btw: why is this a `ref struct`? Is it expected that the struct will hold byreflike fields in the future? Looking at the [current implementation](https://github.com/dotnet/corefx/blob/eb0d4388c339288a589bdd7f5e0c9fcd61da4362/src/System.IO.FileSystem/src/System/IO/RawFindData.cs) there aren't any.